src/engine/sparqlExpressions/DateExpressions.cpp

```diff
@@ -40,10 +40,41 @@ inline auto extractDay = [](std::optional<DateOrLargeYear> d) {
   return Id::makeFromInt(optionalDay.value());
 };
 
+inline auto extractHours = [](std::optional<DateOrLargeYear> d) {
+  if (!d.has_value() || d->getType() != DateOrLargeYear::Type::DateTime) {
+    return Id::makeUndefined();
+  }
+  auto hours = d.value().getDate().getHour();
+  if (hours == -1 ) {
+    return Id::makeUndefined();
+  }
```

Comment:
This check means "There is no time", so you can also check for `hours == -1` in the seconds and the date to consistently get `undefined` for these cases.

Reply:
If you do this, you realize, that the three functions/lambdas are almost the same, just with a `.getHour()`, `.getMinute()` and `.getSeconds()` call. You can template on this as follows:

```
template <auto dateMember, auto makeId> 
inline auto extractTimeComponentImpl = [](std::optional....) {
  // All the canonical sanity checks, is it a date, does it have a time.
  return std::invoke(makeId, std::invoke(dateMember, d.value().getDate()));
}

inline auto extractHours = extractTimeComponentImpl<&date::getHour, &Id::makeFromInt>;
// Two similar lines for minutes and seconds
```

Note: I have written this from the top of my head, it should work, but if you run into trouble, please contact me, or leave it as is.